lib/eknvfs/ekn-vfs.c

```diff
@@ -0,0 +1,340 @@
+{
+  g_return_if_fail (EKN_IS_VFS (self));
+
+  switch (prop_id)
+    {
+      case PROP_SHARDS:
+        ekn_vfs_set_shards (EKN_VFS (self), g_value_get_boxed (value));
+      break;
+      default:
+        G_OBJECT_WARN_INVALID_PROPERTY_ID (self, prop_id, pspec);
+      break;
+    }
+}
+
+static void
+ekn_vfs_get_property (GObject    *self,
+                      guint       prop_id,
+                      GValue     *value,
+                      GParamSpec *pspec)
+{
+  EknVfsPrivate *priv;
+
+  g_return_if_fail (EKN_IS_VFS (self));
+  priv = EKN_VFS_PRIVATE (self);
+
+  switch (prop_id)
+    {
+      case PROP_SHARDS:
+        g_value_set_boxed (value, priv->shards);
+      break;
+      default:
+        G_OBJECT_WARN_INVALID_PROPERTY_ID (self, prop_id, pspec);
+      break;
+    }
+}
+
+static gboolean
+ekn_vfs_is_active (GVfs *self)
+{
+  return TRUE;
+}
+
+static GFile *
+ekn_vfs_get_file_for_path (GVfs *self, const char *path)
+{
+  return g_vfs_get_file_for_path (EKN_VFS_PRIVATE (self)->local, path);
+}
+
+static GFile *
+ekn_vfs_get_file_for_uri (GVfs *self, const char *uri)
+{
+  EknVfsPrivate *priv = EKN_VFS_PRIVATE (self);
+  GFile *retval = NULL;
+  
+  if (uri && g_str_has_prefix (uri, EKN_URI":"))
+    {
+      /* The URI is of the form 'ekn://domain/hash[/resource]' */
+      gchar **tokens = g_strsplit (uri + EKN_SCHEME_LEN, "/", -1);
+
+      if (tokens && tokens[0] && tokens[1])
+        {
+          EosShardRecord *record = NULL;
+          GSList *l;
+
+          /* iterate over all shards until we find a matching record */
+          for (l = priv->shards; l && !record; l = g_slist_next (l))
+            record = eos_shard_shard_file_find_record_by_hex_name (l->data, tokens[1]);
+
+          if (record)
+            {
+              EosShardBlob *blob = record->data;
+
+              /* Use resource, if present */
+              if (tokens[2])
+                blob = eos_shard_record_lookup_blob (record, tokens[2]);
+
+              if (blob)
+                retval = _ekn_file_new (uri, blob);
+            }
+        }
+
+      g_strfreev (tokens);
+    }
+  else if (uri && priv->extensions)
+    {
+      gchar *scheme   = g_uri_parse_scheme (uri);
+      GVfs  *delegate = g_hash_table_lookup (priv->extensions, scheme);
+
+      if (delegate)
+        retval = g_vfs_get_file_for_uri (delegate, uri);
+
+      g_free (scheme);
+    }
+
+  /* This method should never fail */
+  return retval ? retval : g_vfs_get_file_for_uri (priv->local, uri);
+}
+
+static const gchar * const *
+ekn_vfs_get_supported_uri_schemes (GVfs *self)
+{
+  return (const gchar * const *) EKN_VFS_PRIVATE (self)->schemes;
+}
+
+static GFile *
+ekn_vfs_parse_name (GVfs *self, const char *parse_name)
+{
+  if (g_path_is_absolute (parse_name))
+    return g_vfs_parse_name (EKN_VFS_PRIVATE (self)->local, parse_name);
+  else
+    return ekn_vfs_get_file_for_uri (self, parse_name);
+}
+
+static GType
+ekn_vfs_slist_boxed_get_type (void)
```

Comment:
This is not really needed: you can annotate the property using

    /**
     * EknVfs:shards: (element-type GObject)
     * ...

The introspection scanner will be able to tell the bindings what type to expect inside the GList container.